pkg/apply/apply.go

```diff
@@ -3,19 +3,83 @@ package apply
 import (
 	"context"
 	"fmt"
+	corev1 "k8s.io/api/core/v1"
 	"log"
 
 	"github.com/pkg/errors"
 
+	"gopkg.in/yaml.v2"
 	"k8s.io/apimachinery/pkg/api/equality"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/types"
 	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// Find existing object or create if if it doesn't exists
-func findOrCreateObject(ctx context.Context, client k8sclient.Client, obj *uns.Unstructured) (*uns.Unstructured, string, error) {
+type ObjectType int
+
+const (
+	ADDRESSSPOOL ObjectType = iota
+	BGPPEER
+)
+
+func UpdateConfigMapObjs(ctx context.Context, client k8sclient.Client, objs []*uns.Unstructured, objtype ObjectType, namespace string) error {
+	existingStruct, updateStruct := ConfigMapData{}, ConfigMapData{}
+	var configMap corev1.ConfigMap
+
+	if err := client.Get(ctx, k8sclient.ObjectKey{Name: MetalLBConfigMap, Namespace: namespace}, &configMap); err != nil {
+		return nil
+	}
+
+	if err := yaml.Unmarshal([]byte(configMap.Data[MetalLBConfigMap]), &existingStruct); err != nil {
+		return err
+	}
+
+	// Initialize object's field in existing structure to be populated while walking the objects
+	switch objtype {
+	case BGPPEER:
+		existingStruct.Peers = nil
+	case ADDRESSSPOOL:
+		existingStruct.AddressPools = nil
+	}
+
+	for _, obj := range objs {
+		updateStr, ok, err := uns.NestedString(obj.Object, "data", MetalLBConfigMap)
+		if !ok || err != nil {
+			return err
+		}
+		if err := yaml.Unmarshal([]byte(updateStr), &updateStruct); err != nil {
+			return err
+		}
+		switch objtype {
+		case BGPPEER:
+			existingStruct.Peers = append(existingStruct.Peers, updateStruct.Peers...)
+		case ADDRESSSPOOL:
+			existingStruct.AddressPools = append(existingStruct.AddressPools, updateStruct.AddressPools...)
+		}
+	}
+
+	resData, err := yaml.Marshal(existingStruct)
+	if err != nil {
+		return err
+	}
+
+	configMap.Data[MetalLBConfigMap] = string(resData)
+
+	// If there is no more any objects then its safe to delete the configmap
+	if len(existingStruct.AddressPools) == 0 && len(existingStruct.Peers) == 0 {
```

Comment:
can we have a method for this instead having it verbose here? existingStruct.IsEmpty()
I think you can even use
reflect.ValueOf(existingStruct).IsZero() or
reflect.DeepEqual(existingStruct,ConfigMapData{})